File: arc/go/stl/control/control_test.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package control_test

import (
	"context"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	"github.com/synnaxlabs/arc/graph"
	"github.com/synnaxlabs/arc/ir"
	"github.com/synnaxlabs/arc/runtime/node"
	"github.com/synnaxlabs/arc/stl/control"
	. "github.com/synnaxlabs/arc/stl/testutil"
	"github.com/synnaxlabs/arc/symbol"
	. "github.com/synnaxlabs/arc/symbol/testutil"
	"github.com/synnaxlabs/arc/types"
	"github.com/synnaxlabs/x/encoding/msgpack"
	"github.com/synnaxlabs/x/query"
	. "github.com/synnaxlabs/x/testutil"
)

var _ = Describe("Control", func() {
	// setAuthority declares the native's input shape for building test configs.
	setAuthority := NodeSpec{Type: "set_authority", Inputs: types.Params{
		{Name: "value", Type: types.U8()},
		{Name: "channel", Type: types.WriteChan(types.U8())},
	}}
	// constConfig helper builds a set_authority config whose inputs are all
	// consts, compiled from a graph program (covering the graph -> State path).
	constConfig := func(ctx context.Context, value uint8, channel uint32) node.Config {
		return GraphConfig(ctx, NewGraphRoot(nil), "set_authority",
			msgpack.EncodedJSON{"value": value, "channel": channel})
	}

	Describe("NewModule", func() {
		It("Should create factory with state", func(ctx SpecContext) {
			g := graph.Graph{
				Nodes:     []graph.Node{{Key: "set_auth"}},
				Inputs:    map[string]msgpack.EncodedJSON{"set_auth": {"type": "set_authority"}},
				Functions: []ir.Function{{Key: "set_authority"}},
			}
			inter, diagnostics := graph.Analyze(ctx, g, NewGraphRoot(nil))
			Expect(diagnostics.Ok()).To(BeTrue())
			_ = node.New(inter)
			factory := control.NewHost(&control.ProgramState{})
			Expect(factory).ToNot(BeNil())
		})
	})

	Describe("Factory.Create", func() {
		var (
			factory        node.Factory
			s              *node.ProgramState
			authorityState *control.ProgramState
		)
		BeforeEach(func(ctx SpecContext) {
			g := graph.Graph{
				Nodes:     []graph.Node{{Key: "set_auth"}},
				Inputs:    map[string]msgpack.EncodedJSON{"set_auth": {"type": "set_authority"}},
				Functions: []ir.Function{{Key: "set_authority"}},
			}
			analyzed, diagnostics := graph.Analyze(ctx, g, NewGraphRoot(nil))
			Expect(diagnostics.Ok()).To(BeTrue())
			s = node.New(analyzed)
			authorityState = &control.ProgramState{}
			factory = control.NewHost(authorityState)
		})
		It("Should create node for set_authority type", func(ctx SpecContext) {
			constCfg := constConfig(ctx, 200, 42)
			Expect(MustSucceed(factory.Create(ctx, constCfg))).ToNot(BeNil())
			varCfg := setAuthority.Config(NewVarInput[uint8](200), uint32(42))
			Expect(MustSucceed(factory.Create(ctx, varCfg))).ToNot(BeNil())
		})
		It("Should create node for control.set_authority via CompoundFactory", func(ctx SpecContext) {
			compound := node.CompoundFactory{factory}
			cfg := constConfig(ctx, 200, 42)
			cfg.Node.Type = "control.set_authority"
			Expect(MustSucceed(compound.Create(ctx, cfg))).ToNot(BeNil())
		})
		It("Should return NotFound for unknown type", func(ctx SpecContext) {
			cfg := node.Config{
				Node:  ir.Node{Type: "unknown"},
				State: s.Node("set_auth"),
			}
			Expect(factory.Create(ctx, cfg)).Error().To(MatchError(query.ErrNotFound))
		})
		It("Should error when an input value is invalid", func(ctx SpecContext) {
			cfg := node.Config{
				Node: ir.Node{
					Type:   "set_authority",
					Inputs: types.Params{{Name: "value", Type: types.U8(), Value: []any{1}}},
				},
				State: s.Node("set_auth"),
			}
			Expect(factory.Create(ctx, cfg)).Error().To(BeAValidationPathError())
		})
	})

	Describe("Next", func() {
		var (
			authorityState *control.ProgramState
			factory        node.Factory
			outputs        []string
		)
		BeforeEach(func() {
			authorityState = &control.ProgramState{}
			factory = control.NewHost(authorityState)
			outputs = []string{}
		})

		DescribeTable("Should buffer a per-channel authority change",
			func(ctx SpecContext, value, channel any) {
				n := MustSucceed(factory.Create(ctx, setAuthority.Config(value, channel)))
				n.Next(node.Context{Context: ctx, MarkChanged: func(int) {}})
				changes := authorityState.Flush()
				Expect(changes).To(HaveLen(1))
				Expect(changes[0].Authority).To(Equal(uint8(200)))
				Expect(changes[0].Channel).ToNot(BeNil())
				Expect(*changes[0].Channel).To(Equal(uint32(42)))
			},
			Entry("const inputs", uint8(200), uint32(42)),
			Entry("var value", VarOf(uint8(200)), uint32(42)),
		)

		DescribeTable("Should buffer a global authority change",
			func(ctx SpecContext, value, channel any) {
				n := MustSucceed(factory.Create(ctx, setAuthority.Config(value, channel)))
				n.Next(node.Context{Context: ctx, MarkChanged: func(int) {}})
				changes := authorityState.Flush()
				Expect(changes).To(HaveLen(1))
				Expect(changes[0].Authority).To(Equal(uint8(150)))
				Expect(changes[0].Channel).To(BeNil())
			},
			Entry("const inputs", uint8(150), uint32(0)),
			Entry("var value", VarOf(uint8(150)), uint32(0)),
		)

		DescribeTable("Should fire only once before Reset",
			func(ctx SpecContext, value, channel any) {
				n := MustSucceed(factory.Create(ctx, setAuthority.Config(value, channel)))
				nCtx := node.Context{Context: ctx, MarkChanged: func(int) {}}
				n.Next(nCtx)
				n.Next(nCtx)
				n.Next(nCtx)
				changes := authorityState.Flush()
				Expect(changes).To(HaveLen(1))
				Expect(changes[0].Authority).To(Equal(uint8(200)))
			},
			Entry("const inputs", uint8(200), uint32(42)),
			Entry("var value", VarOf(uint8(200)), uint32(42)),
		)

		DescribeTable("Should not call MarkChanged",
			func(ctx SpecContext, value, channel any) {
				n := MustSucceed(factory.Create(ctx, setAuthority.Config(value, channel)))
				n.Next(node.Context{Context: ctx, MarkChanged: func(int) {
					// setAuthority declares no outputs; MarkChanged should never fire.
					outputs = append(outputs, "called")
				}})
				Expect(outputs).To(BeEmpty())
			},
			Entry("const inputs", uint8(200), uint32(42)),
			Entry("var value", VarOf(uint8(200)), uint32(42)),
		)

		It("Should read params through a graph-compiled config", func(ctx SpecContext) {
			n := MustSucceed(factory.Create(ctx, constConfig(ctx, 200, 42)))
			n.Next(node.Context{Context: ctx, MarkChanged: func(int) {}})
			changes := authorityState.Flush()
			Expect(changes).To(HaveLen(1))
			Expect(changes[0].Authority).To(Equal(uint8(200)))
			Expect(*changes[0].Channel).To(Equal(uint32(42)))
		})

		It("Should use the var's declared initial before any write", func(ctx SpecContext) {
			cfg := setAuthority.Config(NewVarInput[uint8](5), uint32(42))
			n := MustSucceed(factory.Create(ctx, cfg))
			n.Next(node.Context{Context: ctx, MarkChanged: func(int) {}})
			changes := authorityState.Flush()
			Expect(changes).To(HaveLen(1))
			Expect(changes[0].Authority).To(Equal(uint8(5)))
		})

		It("Should not re-fire when the var changes without a Reset", func(ctx SpecContext) {
			v := NewVarInput[uint8](1)
			cfg := setAuthority.Config(v, uint32(42))
			v.Set(77)
			n := MustSucceed(factory.Create(ctx, cfg))
			nCtx := node.Context{Context: ctx, MarkChanged: func(int) {}}
			n.Next(nCtx)
			v.Set(33)
			n.Next(nCtx)
			changes := authorityState.Flush()
			Expect(changes).To(HaveLen(1))
			Expect(changes[0].Authority).To(Equal(uint8(77)))
		})

		It("Should read the latest var value on re-fire after Reset", func(ctx SpecContext) {
			v := NewVarInput[uint8](1)
			cfg := setAuthority.Config(v, uint32(42))
			v.Set(77)
			n := MustSucceed(factory.Create(ctx, cfg))
			nCtx := node.Context{Context: ctx, MarkChanged: func(int) {}}
			n.Next(nCtx)
			Expect(authorityState.Flush()[0].Authority).To(Equal(uint8(77)))
			n.Reset()
			v.Set(33)
			n.Next(nCtx)
			changes := authorityState.Flush()
			Expect(changes).To(HaveLen(1))
			Expect(changes[0].Authority).To(Equal(uint8(33)))
		})
	})

	Describe("Reset", func() {
		var (
			authorityState *control.ProgramState
			factory        node.Factory
		)
		BeforeEach(func() {
			authorityState = &control.ProgramState{}
			factory = control.NewHost(authorityState)
		})

		DescribeTable("Should allow re-fire after Reset",
			func(ctx SpecContext, value, channel any) {
				n := MustSucceed(factory.Create(ctx, setAuthority.Config(value, channel)))
				nCtx := node.Context{Context: ctx, MarkChanged: func(int) {}}
				n.Next(nCtx)
				changes := authorityState.Flush()
				Expect(changes).To(HaveLen(1))
				n.Reset()
				n.Next(nCtx)
				changes = authorityState.Flush()
				Expect(changes).To(HaveLen(1))
			},
			Entry("const inputs", uint8(200), uint32(42)),
			Entry("var value", VarOf(uint8(200)), uint32(42)),
		)

		DescribeTable("Should produce same authority on re-fire",
			func(ctx SpecContext, value, channel any) {
				n := MustSucceed(factory.Create(ctx, setAuthority.Config(value, channel)))
				nCtx := node.Context{Context: ctx, MarkChanged: func(int) {}}
				n.Next(nCtx)
				first := authorityState.Flush()
				Expect(first).To(HaveLen(1))
				Expect(first[0].Authority).To(Equal(uint8(200)))
				n.Reset()
				n.Next(nCtx)
				second := authorityState.Flush()
				Expect(second).To(HaveLen(1))
				Expect(second[0].Authority).To(Equal(first[0].Authority))
				Expect(*second[0].Channel).To(Equal(*first[0].Channel))
			},
			Entry("const inputs", uint8(200), uint32(42)),
			Entry("var value", VarOf(uint8(200)), uint32(42)),
		)
	})

	Describe("IsOutputTruthy", func() {
		It("Should always return false", func(ctx SpecContext) {
			factory := control.NewHost(&control.ProgramState{})
			cfg := constConfig(ctx, 200, 42)
			n := MustSucceed(factory.Create(ctx, cfg))
			Expect(n.IsOutputTruthy(0)).To(BeFalse())
			Expect(n.IsOutputTruthy(1)).To(BeFalse())
			Expect(n.IsOutputTruthy(-1)).To(BeFalse())
		})
	})

	Describe("Symbols", func() {
		var root *symbol.Symbol
		BeforeEach(func() { root = symbol.NewRoot(nil, control.NewSymbols()) })
		bare := func(ctx context.Context, name string) *symbol.Symbol {
			return MustSucceed(root.Resolve(ctx, name, symbol.IncludeInternal))
		}
		It("Should expose bare set_authority symbol", func(ctx SpecContext) {
			sym := bare(ctx, "set_authority")
			Expect(sym.Name).To(Equal("set_authority"))
			Expect(sym.Kind).To(Equal(symbol.KindFunction))
		})
		It("Should expose qualified control.set_authority symbol", func(ctx SpecContext) {
			mod := MustSucceed(root.Resolve(ctx, "control", symbol.IncludeInternal))
			sym := MustSucceed(mod.Resolve(ctx, "set_authority", symbol.IncludeInternal))
			Expect(sym.Name).To(Equal("set_authority"))
			Expect(sym.Kind).To(Equal(symbol.KindFunction))
		})
		It("Should declare unified inputs with an activation trigger", func(ctx SpecContext) {
			sym := bare(ctx, "set_authority")
			Expect(sym.Type.Inputs).To(HaveLen(3))
			Expect(sym.Type.Inputs[0].Name).To(Equal(ir.DefaultOutputParam))
			Expect(sym.Type.Inputs[0].Value).To(Equal(uint8(0)))
			Expect(sym.Type.Inputs[1].Name).To(Equal("value"))
			Expect(sym.Type.Inputs[2].Name).To(Equal("channel"))
			Expect(sym.Trigger).To(Equal(symbol.TriggerInput(ir.DefaultOutputParam)))
		})
	})
})

File: arc/cpp/stl/control/control_test.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include "gtest/gtest.h"

#include "x/cpp/mem/indirect.h"
#include "x/cpp/test/test.h"

#include "arc/cpp/ir/ir.h"
#include "arc/cpp/runtime/errors/errors.h"
#include "arc/cpp/runtime/node/factory.h"
#include "arc/cpp/runtime/state/state.h"
#include "arc/cpp/stl/control/control.h"
#include "arc/cpp/stl/testutil/testutil.h"

namespace arc::stl::control {
struct TestSetup {
    ir::IR ir;
    std::shared_ptr<runtime::state::State> state;

    TestSetup(const uint8_t auth_value, const uint32_t channel):
        ir(build_ir(auth_value, channel)),
        state(
            std::make_shared<runtime::state::State>(
                runtime::state::Config{.ir = ir, .channels = {}},
                runtime::errors::noop_handler
            )
        ) {}

    runtime::state::Node make_node() const {
        return ASSERT_NIL_P(this->state->node("set_auth"));
    }

private:
    static ir::IR build_ir(const uint8_t auth_value, const uint32_t channel) {
        types::Param authority_param;
        authority_param.name = "value";
        authority_param.type.kind = types::Kind::U8;
        authority_param.value = auth_value;

        types::Param channel_param;
        channel_param.name = "channel";
        channel_param.type.kind = types::Kind::U32;
        channel_param.value = channel;

        ir::Node ir_node;
        ir_node.key = "set_auth";
        ir_node.type = "set_authority";
        ir_node.inputs.push_back(authority_param);
        ir_node.inputs.push_back(channel_param);

        ir::Function fn;
        fn.key = "test";

        ir::IR ir;
        ir.nodes.push_back(ir_node);
        ir.functions.push_back(fn);
        return ir;
    }
};

runtime::node::Context make_context() {
    return runtime::node::Context{
        .elapsed = x::telem::SECOND,
        .mark_changed = [](size_t) {},
        .report_error = [](const x::errors::Error &) {},
    };
}

/// @brief declares the native's input shape for building test configs.
stl::testutil::NodeSpec set_authority() {
    types::Param value;
    value.name = "value";
    value.type = types::Type{.kind = types::Kind::U8};
    types::Param channel;
    channel.name = "channel";
    channel.type = types::Type{
        .kind = types::Kind::Chan,
        .elem = x::mem::indirect<types::Type>(types::Type{.kind = types::Kind::U8}),
        .chan_direction = types::ChanDirection::Write
    };
    stl::testutil::NodeSpec spec;
    spec.type = "set_authority";
    spec.inputs.push_back(std::move(value));
    spec.inputs.push_back(std::move(channel));
    return spec;
}

/// @brief builds a set_authority config whose inputs are all consts.
stl::testutil::Fixture const_config(const uint8_t value, const uint32_t channel) {
    return set_authority().config({x::json::json(value), x::json::json(channel)});
}

/// @brief returns a state used only to buffer authority changes.
std::shared_ptr<runtime::state::State> auth_state() {
    return std::make_shared<runtime::state::State>(
        runtime::state::Config{.ir = ir::IR{}, .channels = {}},
        runtime::errors::noop_handler
    );
}

TEST(SetAuthorityModuleTest, ReturnsErrorForNullAuthorityValue) {
    TestSetup setup(100, 42);
    auto ir_node = setup.ir.nodes[0];
    for (auto &p: ir_node.inputs)
        if (p.name == "value") p.value = nullptr;

    control::Module module(setup.state);
    ASSERT_OCCURRED_AS_P(
        module.create(runtime::node::Config(setup.ir, ir_node, setup.make_node())),
        x::errors::VALIDATION
    );
}

TEST(SetAuthorityModuleTest, CreatesNode) {
    control::Module module(auth_state());
    auto const_cfg = const_config(200, 42);
    ASSERT_NE(ASSERT_NIL_P(module.create(const_cfg.make_config())), nullptr);
    const std::shared_ptr<stl::testutil::VarBinding>
        v = std::make_shared<stl::testutil::VarInput<uint8_t>>(200);
    auto var_cfg = set_authority().config(
        {v, x::json::json(static_cast<uint32_t>(42))}
    );
    ASSERT_NE(ASSERT_NIL_P(module.create(var_cfg.make_config())), nullptr);
}

TEST(SetAuthorityModuleTest, CreatesNodeWithQualifiedTypeViaMultiFactory) {
    auto f = const_config(200, 42);
    auto ir_node = f.ir_node();
    ir_node.type = "control.set_authority";

    auto module = std::make_shared<control::Module>(auth_state());
    runtime::node::MultiFactory multi({module});
    auto node = ASSERT_NIL_P(
        multi.create(runtime::node::Config(f.program(), ir_node, f.make_node()))
    );
    ASSERT_NE(node, nullptr);
}

TEST(SetAuthorityModuleTest, ReturnsNotFoundForWrongType) {
    TestSetup setup(100, 42);
    auto ir_node = setup.ir.nodes[0];
    ir_node.type = "not_set_authority";

    control::Module module(setup.state);
    ASSERT_OCCURRED_AS_P(
        module.create(runtime::node::Config(setup.ir, ir_node, setup.make_node())),
        x::errors::NOT_FOUND
    );
}

/// @brief runs each spec twice: once with every input a const, once with the value
/// input read live from a variable.
class SetAuthorityInputsTest : public testing::TestWithParam<bool> {
protected:
    std::shared_ptr<runtime::state::State> auth = auth_state();

    stl::testutil::Fixture config(const uint8_t value, const uint32_t channel) const {
        stl::testutil::InputValue v = x::json::json(value);
        if (GetParam())
            v = std::shared_ptr<stl::testutil::VarBinding>(
                stl::testutil::var_of<uint8_t>(value)
            );
        return set_authority().config({v, x::json::json(channel)});
    }
};

INSTANTIATE_TEST_SUITE_P(
    Inputs,
    SetAuthorityInputsTest,
    testing::Values(false, true),
    [](const testing::TestParamInfo<bool> &info) {
        return info.param ? "var_value" : "const_inputs";
    }
);

TEST_P(SetAuthorityInputsTest, BuffersAPerChannelAuthorityChange) {
    auto f = this->config(200, 42);
    control::Module module(this->auth);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    const auto changes = this->auth->flush_authority_changes();
    ASSERT_EQ(changes.size(), 1);
    EXPECT_EQ(changes[0].authority, 200);
    ASSERT_TRUE(changes[0].channel_key.has_value());
    EXPECT_EQ(*changes[0].channel_key, 42);
}

TEST_P(SetAuthorityInputsTest, BuffersAGlobalAuthorityChange) {
    auto f = this->config(150, 0);
    control::Module module(this->auth);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    const auto changes = this->auth->flush_authority_changes();
    ASSERT_EQ(changes.size(), 1);
    EXPECT_EQ(changes[0].authority, 150);
    EXPECT_FALSE(changes[0].channel_key.has_value());
}

TEST_P(SetAuthorityInputsTest, FiresOnlyOnceBeforeReset) {
    auto f = this->config(200, 42);
    control::Module module(this->auth);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    ASSERT_NIL(n->next(ctx));
    ASSERT_NIL(n->next(ctx));
    const auto changes = this->auth->flush_authority_changes();
    ASSERT_EQ(changes.size(), 1);
    EXPECT_EQ(changes[0].authority, 200);
}

TEST_P(SetAuthorityInputsTest, DoesNotCallMarkChanged) {
    auto f = this->config(200, 42);
    control::Module module(this->auth);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    std::vector<std::string> outputs;
    auto ctx = make_context();
    // set_authority declares no outputs; mark_changed should never fire.
    ctx.mark_changed = [&outputs](size_t) { outputs.emplace_back("called"); };
    ASSERT_NIL(n->next(ctx));
    EXPECT_TRUE(outputs.empty());
}

TEST_P(SetAuthorityInputsTest, AllowsReFireAfterReset) {
    auto f = this->config(200, 42);
    control::Module module(this->auth);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    EXPECT_EQ(this->auth->flush_authority_changes().size(), 1);
    n->reset();
    ASSERT_NIL(n->next(ctx));
    EXPECT_EQ(this->auth->flush_authority_changes().size(), 1);
}

TEST_P(SetAuthorityInputsTest, ProducesSameAuthorityOnReFire) {
    auto f = this->config(200, 42);
    control::Module module(this->auth);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    const auto first = this->auth->flush_authority_changes();
    ASSERT_EQ(first.size(), 1);
    EXPECT_EQ(first[0].authority, 200);
    n->reset();
    ASSERT_NIL(n->next(ctx));
    const auto second = this->auth->flush_authority_changes();
    ASSERT_EQ(second.size(), 1);
    EXPECT_EQ(second[0].authority, first[0].authority);
    EXPECT_EQ(*second[0].channel_key, *first[0].channel_key);
}

TEST(SetAuthorityVarTest, UsesTheVarsDeclaredInitialBeforeAnyWrite) {
    const auto auth = auth_state();
    control::Module module(auth);
    const std::shared_ptr<stl::testutil::VarBinding>
        v = std::make_shared<stl::testutil::VarInput<uint8_t>>(5);
    auto f = set_authority().config({v, x::json::json(static_cast<uint32_t>(42))});
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    const auto changes = auth->flush_authority_changes();
    ASSERT_EQ(changes.size(), 1);
    EXPECT_EQ(changes[0].authority, 5);
}

TEST(SetAuthorityVarTest, DoesNotReFireWhenTheVarChangesWithoutAReset) {
    const auto auth = auth_state();
    control::Module module(auth);
    const auto v = std::make_shared<stl::testutil::VarInput<uint8_t>>(1);
    auto f = set_authority().config(
        {std::shared_ptr<stl::testutil::VarBinding>(v),
         x::json::json(static_cast<uint32_t>(42))}
    );
    v->set(77);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    v->set(33);
    ASSERT_NIL(n->next(ctx));
    const auto changes = auth->flush_authority_changes();
    ASSERT_EQ(changes.size(), 1);
    EXPECT_EQ(changes[0].authority, 77);
}

TEST(SetAuthorityVarTest, ReadsTheLatestVarValueOnReFireAfterReset) {
    const auto auth = auth_state();
    control::Module module(auth);
    const auto v = std::make_shared<stl::testutil::VarInput<uint8_t>>(1);
    auto f = set_authority().config(
        {std::shared_ptr<stl::testutil::VarBinding>(v),
         x::json::json(static_cast<uint32_t>(42))}
    );
    v->set(77);
    auto n = ASSERT_NIL_P(module.create(f.make_config()));
    auto ctx = make_context();
    ASSERT_NIL(n->next(ctx));
    EXPECT_EQ(auth->flush_authority_changes()[0].authority, 77);
    n->reset();
    v->set(33);
    ASSERT_NIL(n->next(ctx));
    const auto changes = auth->flush_authority_changes();
    ASSERT_EQ(changes.size(), 1);
    EXPECT_EQ(changes[0].authority, 33);
}
}
